open-metadata-implementation/adapters/open-connectors/governance-daemon-connectors/open-lineage-connectors/janus-connector/pom.xml

```diff
@@ -0,0 +1,107 @@
+    <parent>
+        <artifactId>open-lineage-connectors</artifactId>
+        <groupId>org.odpi.egeria</groupId>
+        <version>1.1-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>janus-connector</artifactId>
+    <dependencies>
+        <dependency>
+            <groupId>org.odpi.egeria</groupId>
+            <artifactId>open-connector-framework</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>org.janusgraph</groupId>
+            <artifactId>janusgraph-core</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>org.apache.tinkerpop</groupId>
+            <artifactId>gremlin-core</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-api</artifactId>
+        </dependency>
+
+<!--        <dependency>-->
+<!--            <groupId>org.janusgraph</groupId>-->
+<!--            <artifactId>janusgraph-cassandra</artifactId>-->
+<!--            <exclusions>-->
+<!--                <exclusion>-->
+<!--                    <artifactId>hibernate-validator</artifactId>-->
+<!--                    <groupId>org.hibernate</groupId>-->
+<!--                </exclusion>-->
+<!--            </exclusions>-->
+<!--        </dependency>-->
+
+        <dependency>
+            <groupId>org.janusgraph</groupId>
+            <artifactId>janusgraph-cql</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>org.janusgraph</groupId>
+            <artifactId>janusgraph-es</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>commons-logging</groupId>
+            <artifactId>commons-logging</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>commons-codec</groupId>
+            <artifactId>commons-codec</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>org.xerial.snappy</groupId>
+            <artifactId>snappy-java</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>org.antlr</groupId>
+            <artifactId>antlr-runtime</artifactId>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-dependency-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <id>analyze</id>
+                        <goals>
+                            <goal>analyze-only</goal>
+                        </goals>
+                        <configuration>
+                            <ignoredUnusedDeclaredDependencies combine.children="append">
+                                <!-- Runtime dependencies for janusgraph-->
+                                <ignoredUnusedDeclaredDependency>org.janusgraph:janusgraph-cql:*</ignoredUnusedDeclaredDependency>
```

Comment:
I may need some help understanding JanusGraph here.

I see that at compile time we are dependent on janusgraph-core and gremlin-core (as per https://docs.janusgraph.org/connecting/java/ ). They are fine (and not excluded). 

At what point are the additional runtime dependencies needed? These are the ones in the exclusion (not marked as 'runtime').  I presume janusgraph-cql, es, cassandra come from the required backend that's been chosen for a *specific* configuration of janus - this is dependent on what the org wants for their janusgraph config -- not a requirement of the connector.

As such I'm wondering if any implementation of janusgraph that we provide should itself be in a different maven artifact - not the connector, and so those dependencies should not be expressed here. Additionally as they are custom it's likely down to the org to configure the libraries appropriately, not us.

Unfortunately I can't easily test my theories without more knowledge, and access to cassandra, but I think we are not adequately seperating the concerns here, and are pulling in more dependencies than required.

I assume the commons logging/codec, snappy, antlr are then needed by the above

Can you clarify any further?

@grahamwallis

Reply:
I updated the dependencies.

Reply:
I see we still have janusgraph es/cql, along with commons-logging & commons-codec

At build time the connector is not dependent on these, so it seems as if these additional dependencies are used at runtime by being found on the classpath.

I am trying to understand (not knowing janus well!) whether this is
 * Essential for anyone using Janus graph in egeria 
OR
 * Is a particular choice of backends (be it for logging or database) for Janus that is specific to your individual deployment option.

If it is the latter I wonder if these do not belong in the connector, but should instead be contributed by the environment, or through an additional packaging construct (like the way that for the server chassis we pop connectors into lib - but do not declare them as dependencies/include their jars)

@grahamwallis can you add anything on this?

I'm just trying to ensure we keep a clear separation

Reply:
Having looked at the janusgraph docs, I see we go with a default of berkeley (which makes sense), so is it correct that by adding the janusgraph components here we are adding in the required runtime jars so that we can do config such as

storage.backend=cql
storage.hostname=localhost

index.search.backend=elasticsearch
index.search.hostname=100.100.101.1, 100.100.101.2
index.search.elasticsearch.client-only=true
I'm presuming We run Janusgraph locally , so the connector is more than a 'connector' .. its effectively the janus implementation too.

So on that basis I can see an argument for including these jars in as dependencies as they are tightly associated with janus and open up some additional configuration options - in this case for cassandra. Does this make sense to those reading? Anyone? ;-)

How about the apache components - can you explain how they are needed at runtime?

Reply:
Based on the backend and the indexing you want to use the dependencies should be loaded dynamically correct. By default we can have Berkley for example. So it is needed to pass the dependencies based on configurations but for now I did not do that so I can speed up the development process since a lot of work is based on the code changes. But you are correct this should be taken into consideration. The Cassandra configuration in my case will be past when you configure the open lineage services. The Apache libraries are needed for janusgraph-es or janusgraph-cql I do not remember which one otherwise the build fails.

Reply:
These properties need to be part of the configurationProperties located in the Connection object for the graph store connector.

Reply:
What I did now is added the properties on the config like this.

`{
	"class": "OpenLineageConfig",
	"openLineageDescription": "Open Lineage Service is used for the storage and querying of lineage",
	"inTopicName": "omas.omas.assetlineage.outTopic",
	"openLineageWiki": "wiki URL",
	"openLineageProvider": "org.odpi.openmetadata.openconnectors.governancedaemonconnectors.openlineageconnectors.janusconnector.JanusConnectorProvider",
	"bufferGraphConfig":{
		"storageBackend": "cql",
		"storageHostname": "127.0.0.1",
		"storageCqlKeyspace": "janusgraph",
		"indexSearchBackend": "elasticsearch",
		"indexSearchHostname": "127.0.0.1:9200"
	}
}`

but I did not do anything for how to load the dependencies based on the implementation someone need to use. How should I do this? S. Now I hardcoded the Cassandra and elastic dependency which is not correct.